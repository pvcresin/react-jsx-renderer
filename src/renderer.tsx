import { Component, createElement, ReactNode } from "react";
import { JSXElement, JSXEvaluator, JSXEvaluatorOption } from "./evaluator";

export interface JSXRendererProps extends Partial<JSXEvaluatorOption> {
  code: string;
  fallback?: (error: Error) => JSX.Element | undefined;
}

interface State {}

export class JSXRenderer extends Component<JSXRendererProps, State> {
  static displayName = 'JSX';

  constructor(props: Readonly<JSXRendererProps>) {
    super(props);

    this.state = {};
  }

  public render() {
    try {
      const chidlren = this.eval(this.props.code);
      return <>{chidlren}</>;
    } catch(err) {
      const error = err instanceof Error ? err : new Error(err.toString());
      return this.props.fallback ? this.props.fallback(error) : <>{error.message}</>;
    }
  }

  private eval(code: string): ReactNode[] {
    const evaluator = new JSXEvaluator({
      binding: this.props.binding || {},
      components: this.props.components || {},
      filters: this.props.filters || [],
    });
    return evaluator.eval(code).map((element) => this.toReactNode(element))
  }

  private toReactNode(element: JSXElement | string | number | boolean): ReactNode {
    if (typeof element === 'object') {
      return createElement(element.component, element.props, element.children.map((child) => this.toReactNode(child)))
    }
    return element;
  }
}
