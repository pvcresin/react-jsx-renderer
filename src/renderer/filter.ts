import { JSXElement, JSXFragment, JSXText } from '../types';

export type JSXElementFilter = (node: JSXElement) => JSXElement | undefined;
export type JSXFragmentFilter = (node: JSXFragment) => JSXFragment | undefined;
export type JSXTextFilter = (node: JSXText) => JSXText | undefined;
