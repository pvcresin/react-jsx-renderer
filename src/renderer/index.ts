export * from './options';
export * from './filter';
export * from './renderer';
