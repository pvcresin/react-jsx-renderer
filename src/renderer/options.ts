import { JSXElementFilter, JSXFragmentFilter, JSXTextFilter } from './filter';

export interface RenderingOptions {
  elementFilters?: JSXElementFilter[];
  fragmentFilters?: JSXFragmentFilter[];
  textFilters?: JSXTextFilter[];

  /**
   * When this option is enabled, non-existent HTML elements will not be rendered.
   */
  disableUnknownHTMLElement?: boolean;
}
