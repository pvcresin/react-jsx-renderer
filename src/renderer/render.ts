import { createElement, Fragment, ReactNode } from 'react';
import { JSXNode, JSXElement, JSXFragment, JSXText } from '../types';
import { RenderingOptions } from './options';

const unknownElementCache = new Map<string, boolean>();

export const render = (node: JSXNode, options: RenderingOptions): ReactNode => {
  if (node === null) return node;
  if (node === undefined) return node;

  switch (typeof node) {
    case 'boolean':
      return node;
    case 'string':
    case 'number':
      return renderJSXText(node, options);
    default:
      return renderJSXNode(node, options);
  }
};

const renderJSXText = (text: JSXText, options: RenderingOptions): ReactNode => {
  return applyFilter(options.textFilters || [], text);
};

const renderJSXNode = (node: JSXElement | JSXFragment, options: RenderingOptions): ReactNode => {
  switch (node.type) {
    case 'element':
      return renderJSXElement(node, options);
    case 'fragment':
      return renderJSXFragment(node, options);
  }
};

const renderJSXElement = (element: JSXElement, options: RenderingOptions): ReactNode => {
  const filtered = applyFilter(options.elementFilters || [], element);
  if (!filtered) return undefined;

  if (options.disableUnknownHTMLElement && typeof filtered.component === 'string') {
    const { component } = filtered;
    if (!unknownElementCache.has(component)) {
      unknownElementCache.set(component, document.createElement(component) instanceof HTMLUnknownElement);
    }
    if (unknownElementCache.get(component)) return undefined;
  }

  return createElement(filtered.component, filtered.props, ...filtered.children.map((child) => render(child, options)));
};

const renderJSXFragment = (fragment: JSXFragment, options: RenderingOptions): ReactNode => {
  const filtered = applyFilter(options.fragmentFilters || [], fragment);

  if (filtered) {
    return createElement(Fragment, filtered.props, ...filtered.children.map((child) => render(child, options)));
  } else {
    return undefined;
  }
};

const applyFilter = <T extends JSXNode>(filters: ((target: T) => T | undefined)[], node: T): T | undefined => {
  return filters.reduce<T | undefined>((prev, filter) => (prev ? filter(prev) : undefined), node);
};
