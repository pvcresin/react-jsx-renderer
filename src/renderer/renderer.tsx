import { ESTree } from 'meriyah';
import { memo, Ref, useEffect, useMemo, VFC } from 'react';
import { evaluateJSX, EvaluateOptions, parse } from '../evaluate';
import { JSXNode } from '../types';
import { RenderingOptions } from './options';
import { renderJSX } from './render';

export type JSXFallbackComponent = VFC<{ error: Error }>;

export interface JSXRendererProps extends EvaluateOptions, RenderingOptions {
  code?: string;
  fallbackComponent?: JSXFallbackComponent;
  refNodes?: Ref<JSXNode[]>;
}

const Renderer: VFC<JSXRendererProps> = (props) => {
  const { code, fallbackComponent, refNodes, ...options } = props;
  const { meriyah, debug } = options;
  const Fallback = fallbackComponent ? fallbackComponent : ({ error }) => <>{error.message}</>;

  try {
    debug && console.group('JSXRenderer');
    useEffect(() => {
      if (typeof refNodes === 'function') {
        refNodes(children || []);
      }
    }, [refNodes]);

    const [program, error] = useMemo(() => {
      let program: ESTree.Program | undefined = undefined;
      let error: Error | undefined = undefined;

      try {
        program = parse(code || '', true, { meriyah, debug });
      } catch (e) {
        error = e instanceof Error ? e : new Error(e);
      }
      return [program, error];
    }, [code, meriyah, debug]);
    const children = program ? evaluateJSX(program, options) : [];

    return (
      <>
        {error && <Fallback error={error} />}
        {children.map((child) => renderJSX(child, options))}
      </>
    );
  } catch (err) {
    const error = err instanceof Error ? err : new Error(err);
    return <Fallback error={error} />;
  } finally {
    debug && console.groupEnd();
  }
};
Renderer.displayName = 'JSXRenderer';

export const JSXRenderer = memo(Renderer);
