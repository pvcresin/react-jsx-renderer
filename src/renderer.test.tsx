import renderer from 'react-test-renderer';
import { JSXRenderer } from './renderer';

describe('JSX', () => {
  it('should render JSX', () => {
    const tree = renderer.create(<JSXRenderer code={`
      <h1>Hello, JSX World</h1>
      <p>This is paragraph.</p>
      <ul className="Hello">
        <li>first</li>
        <li>second</li>
      </ul>
      <>Fragment</>
    `} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('should catch error syntax error', () => {
    const tree = renderer.create(<JSXRenderer code={`<p ...`} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('should catch error with number object', () => {
    const tree = renderer.create(<JSXRenderer code={`<p>{raise()}</p>`} binding={{ raise: () => { throw 1; }}} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

  it('should catch error with object', () => {
    const tree = renderer.create(<JSXRenderer code={`<p>{raise()}</p>`} binding={{ raise: () => { throw { foo: 1 }; }}} />).toJSON();
    expect(tree).toMatchSnapshot();
  })
})
