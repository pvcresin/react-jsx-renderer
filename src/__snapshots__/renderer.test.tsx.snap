// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX should catch error syntax error 1`] = `"Unexpected token (1:9)"`;

exports[`JSX should catch error with number object 1`] = `"1"`;

exports[`JSX should catch error with object 1`] = `"[object Object]"`;

exports[`JSX should render JSX 1`] = `
Array [
  "
      ",
  <h1>
    Hello, JSX World
  </h1>,
  "
      ",
  <p>
    This is paragraph.
  </p>,
  "
      ",
  <ul
    className="Hello"
  >
    
        
    <li>
      first
    </li>
    
        
    <li>
      second
    </li>
    
      
  </ul>,
  "
      ",
  "Fragment",
  "
    ",
]
`;
